src/Microsoft.AspNet.Razor/SourceLocation.cs

```diff
@@ -45,30 +75,49 @@ public override string ToString()
                 CharacterIndex);
         }
 
+        /// <inheritdoc />
         public override bool Equals(object obj)
         {
             return (obj is SourceLocation) && Equals((SourceLocation)obj);
         }
 
+        /// <inheritdoc />
         public override int GetHashCode()
         {
             // LineIndex and CharacterIndex can be calculated from AbsoluteIndex and the document content.
-            return AbsoluteIndex;
+            return HashCodeCombiner.Start().Add(FilePath).Add(AbsoluteIndex).CombinedHash;
         }
 
+        /// <inheritdoc />
         public bool Equals(SourceLocation other)
         {
-            return 
+            return string.Equals(FilePath, other.FilePath, StringComparison.OrdinalIgnoreCase) &&
```

Comment:
Based on a conversation with @loudej yesterday about case sensitivity in paths - this should be `Ordinal` since it's source is programmatic and not something the user specifies.